File: bubble_sort.py
# 冒泡排序

def bubble_sort(array):
    array_length = len(array)
    if array_length <= 1:
        return
    for i in range(array_length):
        for j in range(array_length - i - 1):
            flag = False
            if array[j] > array[j+1]:
                # tmp = array[j]
                # array[j] = array[j+1]
                # array[j+1] = tmp
                array[j], array[j+1] = array[j+1], array[j]
                flag = True
        if not flag:
            break
    
test1 = [2, 3, 4, 1, 6, 4, -1]
test2 = [9, 8, 6, 5, 6, 7, 2, 1, 0]
print(test1, '\n', test2, '\n')
bubble_sort(test1)
bubble_sort(test2)
print(test1, '\n', test2)

File: bubble_sort.c
/* 看完了王争的算法之美相关课程，自己写的冒泡排序*/
#include <stdio.h>
#include <stdlib.h>

void bubbleSort(int a[], int n);

int main()
{
    int n = 10;
    int array[] = { 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 };

    bubbleSort(array, n);

    for (int i = 0; i < n; i++ ) {
        printf("%d\t%d\n", i, array[i]);
    }
    printf("\n");

    system("Pause");
    return 0;
}

void bubbleSort(int a[], int n) {
    if (n <= 1) {
        // 当只有一个元素或没有元素时，无需操作
        return;
    }
    for (int i = 0; i < n; i++) {
        int flag = 0;  // 排序是否发生的标志位，每一次向后走一位要重新置为0
        for(int j = 0; j < n - i - 1; j++) {
            if (a[j] > a[j+1]) {
                int tmp = a[j];
                a[j] = a[j+1];
                a[j+1] = tmp;
                flag = 1;  // 每有一次交换，将flag 置为 1
            }
        }
        if (!flag) {
            break;
        }
    }
}